diff --git a/net/freehaven/tor/control/TorControlConnection.java b/net/freehaven/tor/control/TorControlConnection.java
index 9524612..38b1879 100644
--- a/net/freehaven/tor/control/TorControlConnection.java
+++ b/net/freehaven/tor/control/TorControlConnection.java
@@ -740,7 +740,7 @@ public class TorControlConnection implements TorControlCommands {
      * service with the given hostname (excluding the .onion extension).
      */
     public void forgetHiddenService(String hostname) throws IOException {
-        sendAndWaitForResponse("FORGETHS " + hostname + "\r\n", null);
+        sendAndWaitForResponse("HSFORGET " + hostname + "\r\n", null);
     }
 }
 
