diff -BBur orbot-468076e3/external/Makefile orbot-468076e3-briar/external/Makefile
--- orbot-468076e3af317d65a656d57b1d7b642984ceead6/external/Makefile	2014-01-07 15:47:47.000000000 +0000
+++ orbot-468076e3af317d65a656d57b1d7b642984ceead6-briar/external/Makefile	2014-01-07 14:27:46.000000000 +0000
@@ -18,10 +18,13 @@
 # Android NDK setup:
 NDK_BASE ?= /usr/local/android-ndk
 NDK_PLATFORM_LEVEL ?= 8
-NDK_TOOLCHAIN_VERSION=4.6
+NDK_TOOLCHAIN_VERSION=4.8
 NDK_SYSROOT=$(NDK_BASE)/platforms/android-$(NDK_PLATFORM_LEVEL)/arch-arm
 NDK_UNAME:=`uname -s | tr '[A-Z]' '[a-z]'`
-NDK_PROCESSOR:=`uname -m`
+NDK_PROCESSOR:=$(shell uname -m)
+ifeq ($(NDK_PROCESSOR), i686)
+	NDK_PROCESSOR:=x86
+endif
 
 #NDK_TOOLCHAIN=$(NDK_BASE)/toolchains/arm-linux-androideabi-4.4.3/prebuilt/$(NDK_UNAME)-$(NDK_PROCESSOR)
 NDK_TOOLCHAIN=$(NDK_BASE)/toolchains/arm-linux-androideabi-4.6/prebuilt/$(NDK_UNAME)-$(NDK_PROCESSOR)
