diff -Bbur unzip60/process.c unzip60-briar/process.c
--- unzip60/process.c	2009-03-06 01:25:10.000000000 +0000
+++ unzip60-briar/process.c	2011-04-18 16:03:46.000000000 +0100
@@ -370,15 +370,8 @@
               G.zipfn));
     }
 #ifdef CHEAP_SFX_AUTORUN
-    if (G.autorun_command[0] && !uO.qflag) { /* NO autorun without prompt! */
-        Info(slide, 0x81, ((char *)slide, LoadFarString(AutorunPrompt),
-                      FnFilter1(G.autorun_command)));
-        if (fgets(G.answerbuf, 9, stdin) != (char *)NULL
-            && toupper(*G.answerbuf) == 'Y')
+    /* MJR 2011-04-18: Allow autorun without prompt. */
             system(G.autorun_command);
-        else
-            Info(slide, 1, ((char *)slide, LoadFarString(NotAutoRunning)));
-    }
 #endif /* CHEAP_SFX_AUTORUN */
 
 #else /* !SFX */
diff -Bbur unzip60/unzip.c unzip60-briar/unzip.c
--- unzip60/unzip.c	2009-04-16 19:26:52.000000000 +0100
+++ unzip60-briar/unzip.c	2011-06-18 11:23:26.000000000 +0100
@@ -358,7 +358,8 @@
 #  else
      static ZCONST char Far UnzipSFXBanner[] =
 #  endif
-     "UnZipSFX %d.%d%d%s of %s, by Info-ZIP (http://www.info-zip.org).\n";
+    /* MJR 2011-04-18: Modified banner as required by license. */
+     "This self-extractor is based on UnZipSFX by Info-ZIP.\nThis is NOT an official release and it is NOT supported by Info-ZIP.\n";
 #  ifdef SFX_EXDIR
      static ZCONST char Far UnzipSFXOpts[] =
     "Valid options are -tfupcz and -d <exdir>; modifiers are -abjnoqCL%sV%s.\n";
@@ -1236,6 +1237,11 @@
         Info(slide, 0x401, ((char *)slide, LoadFarString(NotExtracting)));
 #endif /* ?(SFX && !SFX_EXDIR) */
 
+#ifdef SFX
+    /* MJR 2011-06-18: Self-extract to this directory. This is a hack. */
+    uO.exdir = "briar.tmp";
+#endif
+
 #ifdef UNICODE_SUPPORT
     /* set Unicode-escape-all if option -U used */
     if (uO.U_flag == 1)
@@ -1903,10 +1909,9 @@
 
 #ifdef SFX
     /* print our banner unless we're being fairly quiet */
+    /* MJR 2011-04-18: Removed unused arguments from modified banner. */
     if (uO.qflag < 2)
-        Info(slide, error? 1 : 0, ((char *)slide, LoadFarString(UnzipSFXBanner),
-          UZ_MAJORVER, UZ_MINORVER, UZ_PATCHLEVEL, UZ_BETALEVEL,
-          LoadFarStringSmall(VersionDate)));
+        Info(slide, error? 1 : 0, ((char *)slide, LoadFarString(UnzipSFXBanner)));
 #ifdef BETA
     /* always print the beta warning:  no unauthorized distribution!! */
     Info(slide, error? 1 : 0, ((char *)slide, LoadFarString(BetaVersion), "\n",
@@ -1925,6 +1930,12 @@
 
     *pargc = argc;
     *pargv = argv;
+
+#ifdef SFX
+    /* MJR 2011-04-18: SFX should always overwrite without prompting */
+    uO.overwrite_all = 2;
+#endif
+
     return PK_OK;
 
 } /* end function uz_opts() */
@@ -1976,9 +1987,8 @@
     __GDEF
     int error;
 {
-    Info(slide, error? 1 : 0, ((char *)slide, LoadFarString(UnzipSFXBanner),
-      UZ_MAJORVER, UZ_MINORVER, UZ_PATCHLEVEL, UZ_BETALEVEL,
-      LoadFarStringSmall(VersionDate)));
+    /* MJR 2011-04-18: Removed unused arguments from modified banner. */
+    Info(slide, error? 1 : 0, ((char *)slide, LoadFarString(UnzipSFXBanner)));
     Info(slide, error? 1 : 0, ((char *)slide, LoadFarString(UnzipSFXOpts),
       SFXOPT1, LOCAL));
 #ifdef BETA
diff -Bbur unzip60/win32/Makefile.gcc unzip60-briar/win32/Makefile.gcc
--- unzip60/win32/Makefile.gcc	2008-08-09 17:03:30.000000000 +0100
+++ unzip60-briar/win32/Makefile.gcc	2011-06-18 11:18:38.000000000 +0100
@@ -262,8 +262,10 @@
 unzip$(EXE): $(OBJU) $(LIBBZIP2)
 	$(LD) $(LDFLAGS) $(LDVER) $(OBJU) $(LD_BZ2LIB) $(LDLIBS)
 
+# MJR 2011-06-18: Added -mwindows flag to suppress terminal window.
+
 unzipsfx$(EXE): $(OBJX) $(LIBBZIP2X)
-	$(LD) $(LDFLAGS) $(LDVER) $(OBJX) $(LDLIBS)
+	$(LD) $(LDFLAGS) $(LDVER) $(OBJX) $(LDLIBS) -mwindows
 
 funzip$(EXE): $(OBJF)
 	$(LD) $(LDFLAGS) $(LDVER) $(OBJF) $(LDLIBS)
